File: R/zarr_backend.R
#' Zarr Storage Backend
#'
#' @description
#' A storage backend implementation for Zarr array format using the CRAN zarr package.
#' Zarr is a cloud-native array storage format that supports chunked, compressed
#' n-dimensional arrays with concurrent read/write access.
#'
#' @section Experimental:
#' This backend uses the CRAN zarr package which is relatively new (v0.1.1, Dec 2025).
#' It supports Zarr v3 format only - Zarr v2 stores cannot be read.
#' Please report any issues to help improve the package.
#'
#' @details
#' This backend provides efficient access to neuroimaging data stored in Zarr format,
#' which is particularly well-suited for:
#' - Large datasets that don't fit in memory
#' - Cloud storage (S3, GCS, Azure)
#' - Parallel processing workflows
#' - Progressive data access patterns
#'
#' The backend expects Zarr arrays organized as a 4D array with dimensions (x, y, z, time).
#' The CRAN zarr package uses R6 classes and supports Zarr v3 format only.
#'
#' @name zarr-backend
#' @keywords internal
NULL

#' Create a Zarr Backend
#'
#' @description
#' Creates a storage backend for Zarr array data using the CRAN zarr package.
#'
#' @section Experimental:
#' This backend uses the CRAN zarr package which is relatively new (v0.1.1, Dec 2025).
#' It supports Zarr v3 format only - Zarr v2 stores cannot be read.
#' Please report any issues to help improve the package.
#'
#' @param source Character path to Zarr store (directory or URL for remote stores)
#' @param preload Logical, whether to load all data into memory (default: FALSE)
#' @return A zarr_backend S3 object
#' @export
#' @keywords internal
#' @examples
#' \dontrun{
#' # Local Zarr store
#' backend <- zarr_backend("path/to/data.zarr")
#'
#' # Remote store
#' backend <- zarr_backend("https://example.com/data.zarr")
#' }
zarr_backend <- function(source,
                         preload = FALSE) {
  # Validate source first
  if (!is.character(source) || length(source) != 1) {
    stop_fmridataset(
      fmridataset_error_config,
      "source must be a single character string",
      parameter = "source",
      value = class(source)
    )
  }

  # Check if zarr is available
  if (!requireNamespace("zarr", quietly = TRUE)) {
    stop_fmridataset(
      fmridataset_error_config,
      "The zarr package is required for zarr_backend but is not installed.",
      details = "Install with: install.packages('zarr')"
    )
  }

  # Create backend object
  backend <- list(
    source = source,
    preload = preload,
    zarr_array = NULL,
    data_cache = NULL,
    dims = NULL,
    is_open = FALSE
  )

  class(backend) <- c("zarr_backend", "storage_backend")
  backend
}

#' @rdname backend_open
#' @method backend_open zarr_backend
#' @export
backend_open.zarr_backend <- function(backend) {
  if (backend$is_open) {
    return(backend)
  }

  # Open Zarr store
  tryCatch(
    {
      # For local paths, ensure they exist
      if (!grepl("^(https?://|s3://|gs://)", backend$source) &&
        !file.exists(backend$source)) {
        stop_fmridataset(
          fmridataset_error_backend_io,
          sprintf("Zarr store not found: %s", backend$source),
          file = backend$source,
          operation = "open"
        )
      }

      # Open the Zarr array using CRAN zarr package
      backend$zarr_array <- zarr::open_zarr(backend$source)

      # Get array info from ZarrArray R6 object
      array_dims <- backend$zarr_array$shape

      # Validate dimensions (expecting 4D: x, y, z, time)
      if (length(array_dims) != 4) {
        stop_fmridataset(
          fmridataset_error_config,
          sprintf("Expected 4D array, got %dD", length(array_dims)),
          parameter = "source",
          value = backend$source
        )
      }

      # Store dimensions
      backend$dims <- list(
        spatial = array_dims[1:3],
        time = array_dims[4]
      )

      # Preload if requested
      if (backend$preload) {
        message("Preloading Zarr data into memory...")
        backend$data_cache <- backend$zarr_array[, , , , drop = FALSE]
      }

      backend$is_open <- TRUE
    },
    error = function(e) {
      stop_fmridataset(
        fmridataset_error_backend_io,
        sprintf("Failed to open Zarr store: %s", e$message),
        file = backend$source,
        operation = "open"
      )
    }
  )

  backend
}

#' @rdname backend_close
#' @method backend_close zarr_backend
#' @export
backend_close.zarr_backend <- function(backend) {
  # Zarr arrays are stateless, so just clear references
  backend$zarr_array <- NULL
  backend$data_cache <- NULL
  backend$is_open <- FALSE
  invisible(NULL)
}

#' @rdname backend_get_dims
#' @method backend_get_dims zarr_backend
#' @export
backend_get_dims.zarr_backend <- function(backend) {
  if (!backend$is_open) {
    stop_fmridataset(
      fmridataset_error_backend_io,
      "Backend must be opened before accessing dimensions",
      operation = "get_dims"
    )
  }

  backend$dims
}

#' @rdname backend_get_mask
#' @method backend_get_mask zarr_backend
#' @export
backend_get_mask.zarr_backend <- function(backend) {
  if (!backend$is_open) {
    stop_fmridataset(
      fmridataset_error_backend_io,
      "Backend must be opened before accessing mask",
      operation = "get_mask"
    )
  }

  n_voxels <- prod(backend$dims$spatial)

  # Default: all voxels are valid
  # Note: Zarr backend stores single arrays, not separate mask arrays
  # Mask data should be provided externally if needed
  mask <- rep(TRUE, n_voxels)

  # Validate mask
  if (any(is.na(mask))) {
    stop_fmridataset(
      fmridataset_error_backend_io,
      "Mask contains NA values",
      operation = "get_mask"
    )
  }

  if (sum(mask) == 0) {
    stop_fmridataset(
      fmridataset_error_backend_io,
      "Mask has no valid voxels",
      operation = "get_mask"
    )
  }

  mask
}

#' @rdname backend_get_data
#' @method backend_get_data zarr_backend
#' @export
backend_get_data.zarr_backend <- function(backend, rows = NULL, cols = NULL) {
  if (!backend$is_open) {
    stop_fmridataset(
      fmridataset_error_backend_io,
      "Backend must be opened before accessing data",
      operation = "get_data"
    )
  }

  # Get dimensions
  n_timepoints <- backend$dims$time
  n_voxels <- prod(backend$dims$spatial)

  # Default to all rows/cols
  if (is.null(rows)) rows <- seq_len(n_timepoints)
  if (is.null(cols)) cols <- seq_len(n_voxels)

  # Validate indices
  if (any(rows < 1 | rows > n_timepoints)) {
    stop_fmridataset(
      fmridataset_error_config,
      sprintf("Row indices must be between 1 and %d", n_timepoints),
      parameter = "rows"
    )
  }

  if (any(cols < 1 | cols > n_voxels)) {
    stop_fmridataset(
      fmridataset_error_config,
      sprintf("Column indices must be between 1 and %d", n_voxels),
      parameter = "cols"
    )
  }

  # Read data using CRAN zarr package
  if (backend$preload) {
    # Use cached data
    data_4d <- backend$data_cache
  } else {
    # Convert column indices to 3D coordinates
    spatial_dims <- backend$dims$spatial
    coords <- arrayInd(cols, spatial_dims)

    # Determine optimal reading strategy
    # For simplicity, read full array if requesting >50% of data
    proportion <- (length(rows) * length(cols)) / (n_timepoints * n_voxels)

    if (proportion > 0.5) {
      # Read full array
      data_4d <- backend$zarr_array[, , , , drop = FALSE]
    } else {
      # Read full array (zarr subsetting is complex)
      # TODO: Optimize for sparse access patterns
      data_4d <- backend$zarr_array[, , , , drop = FALSE]
    }
  }

  # Reshape to time x voxels matrix
  dim(data_4d) <- c(prod(backend$dims$spatial), backend$dims$time)
  data_matrix <- t(data_4d)

  # Return subset
  data_matrix[rows, cols, drop = FALSE]
}

#' @rdname backend_get_metadata
#' @method backend_get_metadata zarr_backend
#' @export
backend_get_metadata.zarr_backend <- function(backend) {
  if (!backend$is_open) {
    stop_fmridataset(
      fmridataset_error_backend_io,
      "Backend must be opened before accessing metadata",
      operation = "get_metadata"
    )
  }

  metadata <- list()

  # Try to get Zarr attributes from R6 object
  tryCatch(
    {
      if (!is.null(backend$zarr_array$dtype)) {
        metadata$dtype <- backend$zarr_array$dtype
      }
      if (!is.null(backend$zarr_array$chunks)) {
        metadata$chunk_shape <- backend$zarr_array$chunks
      }
    },
    error = function(e) {
      # Attributes not available
    }
  )

  # Add basic info
  metadata$storage_format <- "zarr"
  metadata$zarr_version <- "v3"

  metadata
}

File: python/src/fmridataset/backends/zarr_backend.py
"""Zarr array storage backend.

Port of ``R/zarr_backend.R``.  Requires ``zarr`` (optional dependency).
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

import numpy as np
from numpy.typing import NDArray

from ..backend_protocol import BackendDims, StorageBackend
from ..errors import BackendIOError, ConfigError


class ZarrBackend(StorageBackend):
    """Backend for Zarr-format 4-D arrays.

    Expects the Zarr store to contain a single 4-D array with shape
    ``(x, y, z, time)``.

    Parameters
    ----------
    source : str or Path
        Path to the Zarr store directory.
    mask : ndarray of bool or None
        Optional explicit mask.  If ``None``, all voxels are assumed valid.
    preload : bool
        If True, eagerly load all data into memory.
    """

    def __init__(
        self,
        source: str | Path,
        mask: NDArray[np.bool_] | None = None,
        preload: bool = False,
    ) -> None:
        self._source = Path(source)
        self._external_mask = mask
        self._preload = preload

        self._zarr_array: Any = None
        self._data_cache: NDArray[np.floating[Any]] | None = None
        self._mask_vec: NDArray[np.bool_] | None = None
        self._dims: BackendDims | None = None
        self._is_open = False

    def open(self) -> None:
        try:
            import zarr as zarr_lib
        except ImportError as exc:
            raise ConfigError(
                "zarr is required for ZarrBackend. "
                "Install with: pip install zarr"
            ) from exc

        if not self._source.exists():
            raise BackendIOError(
                f"Zarr store not found: {self._source}",
                file=str(self._source),
                operation="open",
            )

        try:
            self._zarr_array = zarr_lib.open(str(self._source), mode="r")
        except Exception as exc:
            raise BackendIOError(
                f"Failed to open Zarr store: {exc}",
                file=str(self._source),
                operation="open",
            ) from exc

        shape = self._zarr_array.shape
        if len(shape) != 4:
            raise ConfigError(
                f"Expected 4D array, got {len(shape)}D",
                parameter="source",
                value=str(self._source),
            )

        spatial = (int(shape[0]), int(shape[1]), int(shape[2]))
        self._dims = BackendDims(spatial=spatial, time=int(shape[3]))

        n_voxels = int(np.prod(spatial))
        if self._external_mask is not None:
            self._mask_vec = np.asarray(self._external_mask, dtype=np.bool_).ravel()
            if self._mask_vec.size != n_voxels:
                raise ConfigError(
                    f"mask length ({self._mask_vec.size}) != "
                    f"prod(spatial) ({n_voxels})",
                    parameter="mask",
                )
        else:
            self._mask_vec = np.ones(n_voxels, dtype=np.bool_)

        if self._preload:
            raw = np.asarray(self._zarr_array, dtype=np.float64)
            n_time = raw.shape[3]
            flat = raw.reshape(-1, n_time).T  # (time, voxels_flat)
            voxel_idx = np.where(self._mask_vec)[0]
            self._data_cache = flat[:, voxel_idx]

        self._is_open = True

    def close(self) -> None:
        self._zarr_array = None
        self._data_cache = None
        self._is_open = False

    def get_dims(self) -> BackendDims:
        if self._dims is None:
            raise BackendIOError("Backend not opened", operation="get_dims")
        return self._dims

    def get_mask(self) -> NDArray[np.bool_]:
        if self._mask_vec is None:
            raise BackendIOError("Backend not opened", operation="get_mask")
        return self._mask_vec.copy()

    def get_data(
        self,
        rows: NDArray[np.intp] | None = None,
        cols: NDArray[np.intp] | None = None,
    ) -> NDArray[np.floating[Any]]:
        if self._data_cache is not None:
            data = self._data_cache
        else:
            if self._zarr_array is None:
                raise BackendIOError("Backend not opened", operation="get_data")
            raw = np.asarray(self._zarr_array, dtype=np.float64)
            n_time = raw.shape[3]
            flat = raw.reshape(-1, n_time).T
            assert self._mask_vec is not None
            voxel_idx = np.where(self._mask_vec)[0]
            data = flat[:, voxel_idx]

        if rows is not None:
            data = data[rows, :]
        if cols is not None:
            data = data[:, cols]
        return data

    def get_metadata(self) -> dict[str, Any]:
        return {"format": "zarr", "storage_format": "zarr"}
